api/src/main/java/org/apache/iceberg/expressions/Expressions.java

```diff
@@ -123,6 +123,22 @@ public static Expression not(Expression child) {
     return new UnboundPredicate<>(Expression.Operation.NOT_NULL, expr);
   }
 
+  public static <T> UnboundPredicate<T> isNaN(String name) {
+    return new UnboundPredicate<>(Expression.Operation.IS_NAN, ref(name));
+  }
+
+  public static <T> UnboundPredicate<T> isNaN(UnboundTerm<T> expr) {
+    return new UnboundPredicate<>(Expression.Operation.IS_NAN, expr);
+  }
+
+  public static <T> UnboundPredicate<T> notNaN(String name) {
+    return new UnboundPredicate<>(Expression.Operation.NOT_NAN, ref(name));
+  }
+
+  public static <T> UnboundPredicate<T> notNaN(UnboundTerm<T> expr) {
+    return new UnboundPredicate<>(Expression.Operation.NOT_NAN, expr);
+  }
```

Comment:
Do we also need to update the equality predicate to catch `NaN` and rewrite to `isNaN`?

Reply:
I originally thought to update `SparkFilters` to do the rewrite, but this is a much better place. Thanks for the suggestion!

Reply:
I do not fully understand what you mean by "rewrite logic of `or(isNaN, in)`/`and(notNaN, notIn)`" when you talk about rewriting `in`. Can you give some examples of what predicate are you trying to support?

Reply:
So now since we want to handle NaN in `in` predicate, for query `in(1,2, NaN)` to avoid checking for NaN in `in` evaluation all the time we can transform that to `in(1,2) or isNaN`, and `notIn(1,2,NaN)` to `notIn(1, 2) and notNaN`. The problem is where to do that, since `in` and `notIn` are both predicate, and if we are extending them we are transforming a predicate (simpler form) to an expression (complex form), and I think there's no such case in the current code base, and it would touch a lot of existing test cases for this.

Reply:
Okay so it's what I thought, just a bit confused by the notation. 

So for `eq`, what is the benefit of doing it in `UnboundedPredicate` versus just rewriting it in the `Expressions`?

For `in`, I think it is a more complex question.We need to figure out:
1. should syntax like `in(1,2,NaN)` be supported, given it can be written as `is_nan or in(1,2)` on client side
2. if so, `Expressions.in` should return `Expression` as you said, which looks fine to me because the only caller `SparkFilters.convert` also returns an `Expression` in the end.
3. maybe we should tackle this in another PR to keep changes concise.

Reply:
I think it is important to catch problems as early as possible, so I'd rather not add logic to binding if we can do it here. I see what you mean that it would require more test changes, but I think we could avoid that problem by simply rejecting predicates that are passed NaN, just like these methods reject null literals. That's an easy way to get this PR done and not over-complicate binding. What do you think?

Reply:
The main reason for me to suggest rewriting in `UnboundPredicate` is mostly for readability, as `UnboundPredicate` already contains [quite some rewritings](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java#L194-L197), and updating `Expressions` will result in more changes to itself and related test cases. 

I agree with catching problems early, and just to confirm, I guess this only applies to rejecting invalid input, as rewriting equals/notEquals in `Expressions` probably won't help with catching problem as it probably won't throw anything by rewriting without binding? Based on my understanding I will do the following:

1. reject `>/>=/</<= NaN` in `Expressions` (currently in `UnboundPredicate`, will move to `Expressions`)
2. check if `in`/`notIn` literals contain `NaN` and reject
3. rewrite `eq/notEq` in `UnboundPredicate` (already happening)

Please let me know if you have comment!

Reply:
My suggestion to reject invalid predicates should help avoid the problem of rewriting so much test code. What I'm trying to say is that if it requires changing the return type of one of the factory methods, let's throw an `IllegalArgumentException` instead so that we don't need to.

From your list, I agree with 1 and 2, but for 3 I would do the rewrite in Expressions because `isNan` and `notNaN` rewrites will produce `UnboundPredicate` and we don't have to change the tests. Only rewriting `in` and `notIn` would require changing lots of tests, right?

Reply:
Thanks for the quick response! Yeah I think the amount of change to method return type/tests is not a concern now. I just wasn't entirely sure if rewriting `eq` to `isNan` in `Expressions` will help with catching problems early (comparing to rewriting in `UnboundPredicate`), since it seems to me that the related code will not have a chance to throw any exception until `bind()` is called?

Reply:
Yeah, it isn't much earlier in that case. Maybe that actually exposes a problem with rewriting, too.

`Expressions.equal("c", Double.NaN)` if `c` is not a floating point column would result in `isNaN`, which should be rejected while binding expressions. You could argue that it should rewrite to `alwaysFalse` instead following the same logic as `Expressions.equal("intCol", Long.MAX_VALUE)` -- it can't be true.

I think that it would be better to be strict and reject binding in that case because something is clearly wrong. I think a lot of the time, that kind of error would happen when columns are misaligned or predicates are incorrectly converted.

If the result of those errors is just to fail in expression binding, then why rewrite at all? Maybe we should just reject NaN in any predicate and force people to explicitly use `isNaN` and `notNaN`. That way we do throw an exception much earlier in all cases. Plus, we wouldn't have to worry about confusion over whether `NaN` is equal to itself: in Java, a `Double` that holds NaN is equal to itself, but a primitive is not. :confused:

Reply:
Thanks, those are some good points! To make sure I understand correctly/know how to move forward, I have some questions: 

- If I understand correctly, to reject NaN in any predicate sounds like we might go back to the idea of rewriting equals in `SparkFilters` (or in general, the integration point with engines during the query-to-expression translation); or maybe even earlier than that, to let engines to support syntax of `is NaN`? 
- Since to know if a query is eligible to be translated to `isNaN` there has to be some place that ensures the type has to be either double or float, and in iceberg code base we will only know this during binding; are we able to rely on engine to do this check before translating query to `Expression`? 
- And seems like this may only impact `eq` as we decided to do input validation on other `lg/lteq/gt/gteq` and `in` anyway? 
- And if we start to throw exceptions when the code passes in `NaN` to `eq`, that may sound backward incompatible until the engine starts to rewrite NaN?

I guess the conversation is starting to get too detailed, if you wouldn't mind I'll try to follow up on Slack tomorrow and then post the conclusion here?

Reply:
> If I understand correctly, to reject NaN in any predicate sounds like we might go back to the idea of rewriting equals in SparkFilters

Yes. If the engine generally uses `d = NaN` then we can convert that to `isNaN`. But that would be engine-dependent and the Iceberg expression API would not support equals with NaN.

> are we able to rely on engine to do this check before translating query to Expression?

I think so. Most engines will optimize the SQL expressions and handle this already. If not, then it would result in an exception from Iceberg to the user. I think that's okay, too, because as I said above, we want to fail if a NaN is used in an expression with a non-floating-point column, not rewrite to false.

> And seems like this may only impact eq as we decided to do input validation on other lg/lteq/gt/gteq and in anyway?

Yes. This makes all of the handling in `Expressions` consistent: always reject NaN values.

> that may sound backward incompatible until the engine starts to rewrite NaN?

I'm not convinced either way. You could argue that `d = NaN` is ambiguous and that rejecting it is now fixing a bug. That's certainly the case with `d > NaN`, which is not defined. On the other hand, there was _some_ bevhavior before that will now no longer work. So I'd be up for fixing this in Flink and Spark conversions as soon as we can.

Feel free to ping me on Slack!

Reply:
Thank you for the explanation! I think now I understand the full picture. I think I've addressed everything except for rewriting in `SparkFilters` and other engines, which I think this PR is already too big so I'll submit a separate PR for it (likely next week).